plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -356,6 +378,102 @@ private static ConnectorPageSource createOrcPageSource(
+        for (String dereference : path) {
+            List<ColumnIdentity> children = columnIdentity.getChildren();
+            for (int i = 0; i < children.size(); i++) {
+                if (children.get(i).getName().equals(dereference)) {
+                    dereferenceChain.add(i);
+                    columnIdentity = children.get(i);
+                    break;
+                }
+            }
+        }
+
+        return dereferenceChain.build();
+    }
+
+    /**
+     * Creates a mapping between the input {@param columns} and base columns if required.
+     */
+    public static Optional<ReaderColumns> projectColumns(List<IcebergColumnHandle> columns)
+    {
+        // No projection is required if all columns are base columns
+        if (columns.stream().allMatch(IcebergColumnHandle::isBaseColumn)) {
+            return Optional.empty();
+        }
+
+        Map<IcebergColumnHandle, Integer> originalColumnIndexes = Streams.mapWithIndex(columns.stream(), Map::entry)
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> toIntExact(entry.getValue())));
+        // Order the columns in path depth order so existing projections for nested columns can be reused
+        List<IcebergColumnHandle> orderedColumns = originalColumnIndexes.keySet().stream()
+                .sorted(Comparator.comparingInt(col -> col.getPath().size()))
+                .collect(toImmutableList());
+
+        ImmutableList.Builder<ColumnHandle> projectedColumns = ImmutableList.builder();
+        Integer[] outputColumnMapping = new Integer[columns.size()];
+        Map<List<String>, Integer> mappedColumnIndices = new HashMap<>();
+        int projectedColumnIndex = 0;
+
+        for (IcebergColumnHandle column : orderedColumns) {
+            Optional<Integer> existingProjectedColumn = findExistingProjection(mappedColumnIndices, column);
+            int columnIndex = originalColumnIndexes.get(column);
+            if (existingProjectedColumn.isPresent()) {
+                outputColumnMapping[columnIndex] = existingProjectedColumn.get();
+                continue;
+            }
+
+            projectedColumns.add(column);
+            mappedColumnIndices.put(column.getFullPath(), projectedColumnIndex);
+            outputColumnMapping[columnIndex] = projectedColumnIndex;
+            projectedColumnIndex++;
+        }
+
+        return Optional.of(new ReaderColumns(projectedColumns.build(), ImmutableList.copyOf(outputColumnMapping)));
+    }
+
+    private static Optional<Integer> findExistingProjection(Map<List<String>, Integer> mappedColumnIndices, IcebergColumnHandle column)
+    {
+        // TODO: This is n^2 on the number of columns
+        for (Map.Entry<List<String>, Integer> existingMapping : mappedColumnIndices.entrySet()) {
+            List<String> existingPath = existingMapping.getKey();
+            if (column.getPath().size() >= existingPath.size() && column.getPath().subList(0, existingPath.size()).equals(existingPath)) {
```

Comment:
Extract `public static boolean IcebergColumnHandle.isParent(List<String> maybeChild, List<String> maybeParent)` and use where needed. Also as an implementation of `ICH.isParentOf()`

Reply:
Or maybe what will be enough is to extract `ICH.isParentOf(List<String> path)` / `ICH.isChildOf(List<String> path` or both.